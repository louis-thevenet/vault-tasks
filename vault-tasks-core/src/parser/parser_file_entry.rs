use std::iter::Peekable;

use color_eyre::{eyre::bail, Result};
use tracing::{debug, error};
use winnow::{
    ascii::{space0, space1},
    combinator::{alt, preceded, repeat},
    token::{take_till, take_while},
    PResult, Parser,
};

use crate::{task::Task, vault_data::VaultData, TasksConfig};

use super::task::parse_task;

enum FileToken {
    /// Name, Heading level
    Header((String, usize)),
    /// Content, Indent length
    Description(String, usize),
    /// Task, Indent length
    Task(Task, usize),
    /// A tag found outside a task in the file
    FileTag(String),
}

#[allow(clippy::module_name_repetitions)]
pub struct ParserFileEntry<'a> {
    pub config: &'a TasksConfig,
    pub filename: String,
}

impl<'i> ParserFileEntry<'i> {
    fn parse_indent(input: &mut &str) -> PResult<usize> {
        let indent_length: String = repeat(1.., " ").parse_next(input)?;
        Ok(indent_length.len())
    }
    fn parse_task(&self, input: &mut &str) -> PResult<FileToken> {
        let indent_length = Self::parse_indent(input).unwrap_or(0);

        let mut task_parser =
            |input: &mut &str| parse_task(input, self.filename.clone(), self.config);
        let task_res = task_parser.parse_next(input)?;
        Ok(FileToken::Task(task_res, indent_length))
    }
    fn parse_header(input: &mut &str) -> PResult<FileToken> {
        let header_depth: String = repeat(1.., "#").parse_next(input)?;
        let header_content = preceded(space0, take_till(1.., |c| c == '\n')).parse_next(input)?;

        Ok(FileToken::Header((
            header_content.to_string(),
            header_depth.len(),
        )))
    }
    fn parse_description(input: &mut &str) -> PResult<FileToken> {
        let indent_length = space1.map(|s: &str| s.len()).parse_next(input)?;
        let desc_content = take_till(1.., |c| c == '\n').parse_next(input)?;
        Ok(FileToken::Description(
            desc_content.to_string(),
            indent_length,
        ))
    }
    fn parse_file_tag(input: &mut &str) -> PResult<FileToken> {
        let tag = preceded(
            '#',
            take_while(1.., ('_', '0'..='9', 'A'..='Z', 'a'..='z', '0'..='9')),
        )
        .parse_next(input)?;
        Ok(FileToken::FileTag(tag.to_owned()))
    }
    fn insert_task_at(
        file_entry: &mut VaultData,
        task: Task,
        header_depth: usize,
        indent_length: usize,
    ) -> Result<()> {
        fn append_task_aux(
            file_entry: &mut VaultData,
            task_to_insert: Task,
            current_header_depth: usize,
            target_header_depth: usize,
            current_task_depth: usize,
            target_task_depth: usize,
        ) -> Result<()> {
            match file_entry {
                VaultData::Header(_, _, header_children) => {
                    match current_header_depth.cmp(&target_header_depth) {
                        std::cmp::Ordering::Greater => panic!("Target header level was greater than current level which is impossible"), // shouldn't happen
                        std::cmp::Ordering::Equal => {
                            // Found correct header level
                            if current_task_depth == target_task_depth {
                                header_children.push(VaultData::Task(task_to_insert));
                                Ok(())
                            } else {
                                for child in header_children.iter_mut().rev() {
                                    if let VaultData::Task(_task) = child {
                                        return append_task_aux(
                                            child,
                                            task_to_insert,
                                            current_header_depth,
                                            target_header_depth,
                                            current_task_depth + 1,
                                            target_task_depth,
                                        );
                                    }
                                }
                                bail!("Cound't find correct parent task to insert task {}", task_to_insert.name)
                            }
                        }
                        std::cmp::Ordering::Less => {
                            // Going deeper in header levels
                            for child in header_children.iter_mut().rev() {
                                if let VaultData::Header(_,_, _) = child {
                                    return append_task_aux(
                                        child,
                                        task_to_insert,
                                        current_header_depth + 1,
                                        target_header_depth,
                                        current_task_depth,
                                        target_task_depth,
                                    );
                                }
                            }
                                bail!("Cound't find correct parent header to insert task {}", task_to_insert.name)
                        }
                    }
                }
                VaultData::Task(task) => {
                    let mut current_task_depth = current_task_depth;
                    let mut last_task = task;
                    while current_task_depth < target_task_depth {
                        if last_task.subtasks.is_empty() {
                            error!("Could not find parent task, indenting may be wrong. Closest task line number: {}",last_task.line_number);
                            bail!("Failed to insert task")
                        }
                        last_task = last_task.subtasks.last_mut().unwrap();
                        current_task_depth += 1;
                    }
                    last_task.subtasks.push(task_to_insert);
                    Ok(())
                }
                VaultData::Directory(_, _) => {
                    bail!("Failed to insert task: tried to insert into a directory")
                }
            }
        }
        append_task_aux(file_entry, task, 0, header_depth, 0, indent_length)
    }
    /// Inserts a `FileEntry` at the specific header `depth` in `file_entry`.
    fn insert_header_at(
        file_entry: &mut VaultData,
        object: VaultData,
        target_header_depth: usize,
        target_task_depth: usize,
    ) {
        fn insert_at_aux(
            file_entry: &mut VaultData,
            object: VaultData,
            current_header_depth: usize,
            target_header_depth: usize,
            current_task_depth: usize,
            target_task_depth: usize,
        ) {
            match file_entry {
                VaultData::Header(_, _, header_children) => {
                    match (current_header_depth).cmp(&target_header_depth) {
                        std::cmp::Ordering::Greater => error!(
                            "bad call to `insert_at`, file_entry:{file_entry}\nobject:{object}"
                        ), // shouldn't happen
                        std::cmp::Ordering::Equal => {
                            // Found correct header level
                            if current_task_depth == target_task_depth {
                                header_children.push(object);
                            } else {
                                for child in header_children.iter_mut().rev() {
                                    if let VaultData::Task(_) = child {
                                        return insert_at_aux(
                                            child,
                                            object,
                                            current_header_depth,
                                            target_header_depth,
                                            current_task_depth + 1,
                                            target_task_depth,
                                        );
                                    }
                                }
                            }
                        }
                        std::cmp::Ordering::Less => {
                            // Still haven't found correct header level, going deeper
                            for child in header_children.iter_mut().rev() {
                                if let VaultData::Header(_, _, _) = child {
                                    insert_at_aux(
                                        child,
                                        object,
                                        current_header_depth + 1,
                                        target_header_depth,
                                        current_task_depth,
                                        target_task_depth,
                                    );
                                    return;
                                }
                            }
                            header_children.push(object);
                        }
                    }
                }
                VaultData::Task(_) => {
                    error!("Error: tried to insert a header into a task");
                }
                VaultData::Directory(name, _) => {
                    error!("Error: tried to insert a header into a directory : {name}");
                }
            }
        }
        insert_at_aux(
            file_entry,
            object,
            0,
            target_header_depth,
            0,
            target_task_depth,
        );
    }

    /// Appends `desc` to the description of an existing `Task` in the `FileEntry`.
    fn append_description(
        file_entry: &mut VaultData,
        desc: String,
        target_header_depth: usize,
        target_task_depth: usize,
    ) -> Result<()> {
        fn append_description_aux(
            file_entry: &mut VaultData,
            desc: String,
            current_header_depth: usize,
            target_header_depth: usize,
            current_task_depth: usize,
            target_task_depth: usize,
        ) -> Result<()> {
            match file_entry {
                VaultData::Header(_, _, header_children) => {
                    match current_header_depth.cmp(&target_header_depth) {
                        std::cmp::Ordering::Greater => panic!("bad call for desc"), // shouldn't happen
                        std::cmp::Ordering::Equal => {
                            // Found correct header level
                            for child in header_children.iter_mut().rev() {
                                if let VaultData::Task(mut task) = child.clone() {
                                    if current_task_depth == target_task_depth {
                                        match &mut task.description {
                                            Some(d) => {
                                                d.push('\n');
                                                d.push_str(&desc.clone());
                                            }
                                            None => task.description = Some(desc.clone()),
                                        }
                                        *child = VaultData::Task(task);
                                    } else {
                                        return append_description_aux(
                                            child,
                                            desc,
                                            current_header_depth,
                                            target_header_depth,
                                            current_task_depth + 1,
                                            target_task_depth,
                                        );
                                    }
                                }
                            }
                            Ok(())
                        }
                        std::cmp::Ordering::Less => {
                            // Going deeper in header levels
                            for child in header_children.iter_mut().rev() {
                                if let VaultData::Header(_, _, _) = child {
                                    return append_description_aux(
                                        child,
                                        desc,
                                        current_header_depth + 1,
                                        target_header_depth,
                                        current_task_depth,
                                        target_task_depth,
                                    );
                                }
                            }
                            bail!("Failed to insert description: previous task not found");
                        }
                    }
                }
                VaultData::Task(task) => {
                    fn insert_desc_task(
                        description: String,
                        task: &mut Task,
                        current_level: usize,
                        target_level: usize,
                    ) -> Result<()> {
                        if current_level == target_level {
                            match &mut task.description {
                                Some(d) => {
                                    d.push('\n');
                                    d.push_str(&description);
                                    Ok(())
                                }
                                None => {
                                    task.description = Some(description.clone());
                                    Ok(())
                                }
                            }
                        } else if let Some(task) = task.subtasks.last_mut() {
                            insert_desc_task(description, task, current_level + 1, target_level)
                        } else {
                            debug!("Description was too indented, adding to closest task: {description}");
                            insert_desc_task(description, task, current_level + 1, target_level)
                        }
                    }
                    insert_desc_task(desc, task, current_task_depth, target_task_depth)
                }
                VaultData::Directory(_, _) => {
                    bail!("Failed to insert description: tried to insert into a directory")
                }
            }
        }
        append_description_aux(
            file_entry,
            desc,
            0,
            target_header_depth,
            0,
            target_task_depth,
        )
    }

    /// Recursively parses the input file passed as a string.
    fn parse_file_aux<'a, I>(
        &self,
        mut input: Peekable<I>,
        file_entry: &mut VaultData,
        file_tags: &mut Vec<String>,
        header_depth: usize,
    ) where
        I: Iterator<Item = (usize, &'a str)>,
    {
        let mut parser = alt((
            Self::parse_file_tag,
            Self::parse_header,
            |input: &mut &str| self.parse_task(input),
            Self::parse_description,
        ));

        let line_opt = input.next();
        if line_opt.is_none() {
            return;
        }

        let (line_number, mut line) = line_opt.unwrap();

        match parser.parse_next(&mut line) {
            Ok(FileToken::Task(mut task, indent_length)) => {
                task.line_number = line_number + 1; // line 1 was element 0 of iterator
                if Self::insert_task_at(
                    file_entry,
                    task,
                    header_depth,
                    indent_length / self.config.indent_length,
                )
                .is_err()
                {
                    error!("Failed to insert task");
                }
                self.parse_file_aux(input, file_entry, file_tags, header_depth);
            }
            Ok(FileToken::Header((header, new_depth))) => {
                Self::insert_header_at(
                    file_entry,
                    VaultData::Header(new_depth, header, vec![]),
                    new_depth - 1,
                    0,
                );
                self.parse_file_aux(input, file_entry, file_tags, new_depth);
            }
            Ok(FileToken::Description(description, indent_length)) => {
                if Self::append_description(
                    file_entry,
                    description.clone(),
                    header_depth,
                    indent_length / self.config.indent_length,
                )
                .is_err()
                {
                    error!("Failed to insert description {description}");
                }
                self.parse_file_aux(input, file_entry, file_tags, header_depth);
            }
            Ok(FileToken::FileTag(tag)) => {
                if !file_tags.contains(&tag) {
                    file_tags.push(tag);
                }
                self.parse_file_aux(input, file_entry, file_tags, header_depth);
            }
            Err(_) => self.parse_file_aux(input, file_entry, file_tags, header_depth),
        }
    }

    /// Removes any empty headers from a `FileEntry`
    fn clean_file_entry(file_entry: &mut VaultData) -> Option<&VaultData> {
        match file_entry {
            VaultData::Header(_, _, children) | VaultData::Directory(_, children) => {
                let mut actual_children = vec![];
                for child in children.iter_mut() {
                    let mut child_clone = child.clone();
                    if Self::clean_file_entry(&mut child_clone).is_some() {
                        actual_children.push(child_clone);
                    }
                }
                *children = actual_children;
                if children.is_empty() {
                    None
                } else {
                    Some(file_entry)
                }
            }
            VaultData::Task(_) => Some(file_entry),
        }
    }

    pub fn parse_file(&mut self, filename: &str, input: &&str) -> Option<VaultData> {
        let lines = input.split('\n');

        let mut res = VaultData::Header(0, filename.to_owned(), vec![]);
        let mut file_tags = vec![];
        self.filename = filename.to_string();
        self.parse_file_aux(lines.enumerate().peekable(), &mut res, &mut file_tags, 0);

        if self.config.file_tags_propagation {
            file_tags.iter().for_each(|t| add_global_tag(&mut res, t));
        }

        // Filename is changed from Header to Directory variant at the end
        if let Some(VaultData::Header(_, name, children)) = Self::clean_file_entry(&mut res) {
            Some(VaultData::Directory(name.clone(), children.clone()))
        } else {
            None
        }
    }
}

fn add_global_tag(file_entry: &mut VaultData, tag: &String) {
    fn add_tag_aux(file_entry: &mut VaultData, tag: &String) {
        match file_entry {
            VaultData::Header(_, _, children) | VaultData::Directory(_, children) => {
                for child in children.iter_mut().rev() {
                    add_tag_aux(child, tag);
                }
            }
            VaultData::Task(task) => {
                fn insert_tag_task(task: &mut Task, tag: &String) {
                    match task.tags.clone() {
                        Some(mut tags) if !tags.contains(tag) => {
                            tags.push(tag.to_string());
                            task.tags = Some(tags);
                        }
                        None => task.tags = Some(vec![tag.to_string()]),
                        _ => (),
                    }

                    for st in &mut task.subtasks {
                        insert_tag_task(st, tag);
                    }
                }
                insert_tag_task(task, tag);
            }
        }
    }
    add_tag_aux(file_entry, tag);
}
#[cfg(test)]
mod tests {

    use insta::assert_snapshot;

    use super::ParserFileEntry;

    use crate::{
        parser::parser_file_entry::add_global_tag, task::Task, vault_data::VaultData, TasksConfig,
    };
    #[test]
    fn test_with_useless_headers() {
        let input = r"# 1 useless
## 2 useless
### 3 useless

# 2 useful
### 3 useless
## 4 useful
- [ ] test
  test
  desc
"
        .split('\n')
        .enumerate()
        .peekable();

        let config = TasksConfig {
            indent_length: 2,
            ..Default::default()
        };
        let mut res = VaultData::Header(0, "Test".to_string(), vec![]);
        let parser = ParserFileEntry {
            config: &config,
            filename: String::new(),
        };
        let expected = VaultData::Header(
            0,
            "Test".to_string(),
            vec![
                VaultData::Header(
                    1,
                    "1 useless".to_string(),
                    vec![VaultData::Header(
                        2,
                        "2 useless".to_string(),
                        vec![VaultData::Header(3, "3 useless".to_string(), vec![])],
                    )],
                ),
                VaultData::Header(
                    1,
                    "2 useful".to_string(),
                    vec![
                        VaultData::Header(3, "3 useless".to_string(), vec![]),
                        VaultData::Header(
                            2,
                            "4 useful".to_string(),
                            vec![VaultData::Task(Task {
                                name: "test".to_string(),
                                line_number: 8,
                                description: Some("test\ndesc".to_string()),
                                ..Default::default()
                            })],
                        ),
                    ],
                ),
            ],
        );
        parser.parse_file_aux(input, &mut res, &mut vec![], 0);
        assert_eq!(res, expected);

        let expected_after_cleaning = VaultData::Header(
            0,
            "Test".to_string(),
            vec![VaultData::Header(
                1,
                "2 useful".to_string(),
                vec![VaultData::Header(
                    2,
                    "4 useful".to_string(),
                    vec![VaultData::Task(Task {
                        name: "test".to_string(),
                        line_number: 8,
                        description: Some("test\ndesc".to_string()),
                        ..Default::default()
                    })],
                )],
            )],
        );
        ParserFileEntry::clean_file_entry(&mut res);
        assert_eq!(res, expected_after_cleaning);
    }
    #[test]
    fn test_simple_input() {
        let input = r"# 1 Header
- [ ] Task

## 2 Header
### 3 Header
- [ ] Task
- [ ] Task 2
## 2 Header 2
- [ ] Task
  Description

"
        .split('\n')
        .enumerate()
        .peekable();

        let config = TasksConfig {
            indent_length: 2,
            ..Default::default()
        };
        let mut res = VaultData::Header(0, "Test".to_string(), vec![]);
        let parser = ParserFileEntry {
            config: &config,
            filename: String::new(),
        };
        let expected = VaultData::Header(
            0,
            "Test".to_string(),
            vec![VaultData::Header(
                1,
                "1 Header".to_string(),
                vec![
                    VaultData::Task(Task {
                        name: "Task".to_string(),
                        line_number: 2,
                        ..Default::default()
                    }),
                    VaultData::Header(
                        2,
                        "2 Header".to_string(),
                        vec![VaultData::Header(
                            3,
                            "3 Header".to_string(),
                            vec![
                                VaultData::Task(Task {
                                    name: "Task".to_string(),
                                    line_number: 6,
                                    ..Default::default()
                                }),
                                VaultData::Task(Task {
                                    name: "Task 2".to_string(),
                                    line_number: 7,
                                    ..Default::default()
                                }),
                            ],
                        )],
                    ),
                    VaultData::Header(
                        2,
                        "2 Header 2".to_string(),
                        vec![VaultData::Task(Task {
                            name: "Task".to_string(),
                            line_number: 9,
                            description: Some("Description".to_string()),
                            ..Default::default()
                        })],
                    ),
                ],
            )],
        );
        parser.parse_file_aux(input, &mut res, &mut vec![], 0);
        assert_eq!(res, expected);
    }
    #[test]
    fn test_insert_global_tag() {
        let input = r"# 1 Header
- [ ] Task

## 2 Header
### 3 Header
- [ ] Task
- [ ] Task 2
## 2 Header 2
- [ ] Task
  Description

"
        .split('\n')
        .enumerate()
        .peekable();

        let config = TasksConfig {
            indent_length: 2,
            ..Default::default()
        };
        let mut res = VaultData::Header(0, "Test".to_string(), vec![]);
        let parser = ParserFileEntry {
            config: &config,
            filename: String::new(),
        };
        parser.parse_file_aux(input, &mut res, &mut vec![], 0);
        add_global_tag(&mut res, &String::from("test"));
        assert_snapshot!(res);
    }
    #[test]
    fn test_fake_description() {
        let input = r"# 1 Header
  test
- [ ] Task

## 2 Header
  test
"
        .split('\n')
        .enumerate()
        .peekable();

        let config = TasksConfig {
            indent_length: 2,
            ..Default::default()
        };
        let mut res = VaultData::Header(0, "Test".to_string(), vec![]);
        let parser = ParserFileEntry {
            config: &config,
            filename: String::new(),
        };
        let expected = VaultData::Header(
            0,
            "Test".to_string(),
            vec![VaultData::Header(
                1,
                "1 Header".to_string(),
                vec![
                    VaultData::Task(Task {
                        name: "Task".to_string(),
                        line_number: 3,
                        ..Default::default()
                    }),
                    VaultData::Header(2, "2 Header".to_string(), vec![]),
                ],
            )],
        );
        parser.parse_file_aux(input, &mut res, &mut vec![], 0);
        assert_eq!(res, expected);
    }
    #[test]
    fn test_nested_tasks() {
        let input = r"# 1 Header
## Test
- [ ] Test a
  - [ ] Test b
    - [ ] Test c
"
        .split('\n')
        .enumerate()
        .peekable();

        let config = TasksConfig {
            indent_length: 2,
            ..Default::default()
        };
        let mut res = VaultData::Header(0, "Test".to_string(), vec![]);
        let parser = ParserFileEntry {
            config: &config,
            filename: String::new(),
        };
        let expected = VaultData::Header(
            0,
            "Test".to_string(),
            vec![VaultData::Header(
                1,
                "1 Header".to_string(),
                vec![VaultData::Header(
                    2,
                    "Test".to_string(),
                    vec![VaultData::Task(Task {
                        name: "Test a".to_string(),
                        line_number: 3,
                        subtasks: vec![Task {
                            name: "Test b".to_string(),
                            line_number: 4,
                            subtasks: vec![Task {
                                name: "Test c".to_string(),
                                line_number: 5,
                                ..Default::default()
                            }],
                            ..Default::default()
                        }],
                        ..Default::default()
                    })],
                )],
            )],
        );
        parser.parse_file_aux(input, &mut res, &mut vec![], 0);
        println!("{res:#?}");
        assert_eq!(res, expected);
    }
    #[test]
    fn test_nested_tasks_desc() {
        let input = r"# 1 Header
- [ ] t1
  t1
  - [ ] t2
    t2
  t1
    t2
    - [ ] t3
    t2
  t1
      t3
  t1
      - [ ] t4
    t2
        t4
      t3
        t4

"
        .split('\n')
        .enumerate()
        .peekable();

        let config = TasksConfig {
            indent_length: 2,
            ..Default::default()
        };
        let mut res = VaultData::Header(0, "Test".to_string(), vec![]);
        let parser = ParserFileEntry {
            config: &config,
            filename: String::new(),
        };
        parser.parse_file_aux(input, &mut res, &mut vec![], 0);
        assert_snapshot!(res);
    }
}
