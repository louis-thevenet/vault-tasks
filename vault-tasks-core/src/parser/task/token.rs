use chrono::{NaiveDate, NaiveTime};

use crate::task::State;

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum Token {
    DueDate(NaiveDate),
    DueTime(NaiveTime),
    Name(String),
    Priority(usize),
    Tag(String),
    State(State),
    TodayFlag,
}
