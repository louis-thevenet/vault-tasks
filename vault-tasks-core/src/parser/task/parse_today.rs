use winnow::{
    combinator::{alt, preceded},
    PResult, Parser,
};

use super::token::Token;

/// Parses a `Token::TodayFlag` of the form of the form "@t", @tdy", "@tod" or "@today".
pub fn parse_today(input: &mut &str) -> PResult<Token> {
    preceded('@', alt(("today", "tod", "tdy", "t"))).parse_next(input)?;
    Ok(Token::TodayFlag)
}

#[cfg(test)]
mod tests {
    use crate::parser::task::{parse_today::parse_today, token::Token};

    #[test]
    fn test_parse_today_tag() {
        let mut with_today = "@t";
        assert_eq!(parse_today(&mut with_today), Ok(Token::TodayFlag));
        let mut with_today = "@today";
        assert_eq!(parse_today(&mut with_today), Ok(Token::TodayFlag));
        let mut with_today = "@tdy";
        assert_eq!(parse_today(&mut with_today), Ok(Token::TodayFlag));
    }
    #[test]
    fn test_parse_today_tag_fail() {
        let mut should_fail = "today";
        assert!(parse_today(&mut should_fail).is_err());
    }
}
