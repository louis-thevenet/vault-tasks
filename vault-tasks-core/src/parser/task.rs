mod parse_today;
mod parser_due_date;
mod parser_priorities;
mod parser_state;
mod parser_tags;
mod parser_time;
mod token;

use chrono::NaiveDateTime;
use parse_today::parse_today;
use parser_due_date::parse_naive_date;
use parser_priorities::parse_priority;
use parser_state::parse_task_state;
use parser_tags::parse_tag;
use parser_time::parse_naive_time;
use token::Token;
use tracing::error;
use winnow::{
    combinator::{alt, fail, repeat},
    token::any,
    PResult, Parser,
};

use crate::{
    task::{DueDate, Task},
    TasksConfig,
};

/// Parses a `Token` from an input string.FileEntry
fn parse_token(input: &mut &str, config: &TasksConfig) -> PResult<Token> {
    alt((
        |input: &mut &str| parse_naive_date(input, config.use_american_format),
        parse_naive_time,
        parse_tag,
        parse_task_state,
        parse_priority,
        parse_today,
        |input: &mut &str| {
            let res = repeat(0.., any)
                .fold(String::new, |mut string, c| {
                    string.push(c);
                    string
                })
                .parse_next(input)?;
            Ok(Token::Name(res))
        },
    ))
    .parse_next(input)
}

/// Parses a `Task` from an input string. Filename must be specified to be added to the task.
///
/// # Errors
///
/// Will return an error if the task can't be parsed.
#[allow(clippy::module_name_repetitions)]
pub fn parse_task(input: &mut &str, filename: String, config: &TasksConfig) -> PResult<Task> {
    let task_state = match parse_task_state(input)? {
        Token::State(state) => Ok(state),
        _ => fail(input),
    }?;

    let mut token_parser = |input: &mut &str| parse_token(input, config);

    let tokens = input
        .split_ascii_whitespace()
        .map(|token| token_parser.parse(token));

    let mut task = Task {
        state: task_state,
        filename,
        ..Default::default()
    };

    // Placeholders for a date and a time
    let mut due_date_opt = None;
    let mut due_time_opt = None;
    let mut name_vec = vec![]; // collects words that aren't tokens from the input string

    for token_res in tokens {
        match token_res {
            Ok(Token::DueDate(date)) => due_date_opt = Some(date),
            Ok(Token::DueTime(time)) => due_time_opt = Some(time),
            Ok(Token::Name(name)) => name_vec.push(name),
            Ok(Token::Priority(p)) => task.priority = p,
            Ok(Token::State(state)) => task.state = state,
            Ok(Token::Tag(tag)) => {
                if let Some(ref mut tags) = task.tags {
                    tags.push(tag);
                } else {
                    task.tags = Some(vec![tag]);
                }
            }
            Ok(Token::TodayFlag) => task.is_today = true,
            Err(error) => error!("Error: {error:?}"),
        }
    }

    if !name_vec.is_empty() {
        task.name = name_vec.join(" ");
    }

    let now = chrono::Local::now();
    let (due_date, has_date) = (
        due_date_opt.unwrap_or_else(|| now.date_naive()),
        due_date_opt.is_some(),
    );
    let (due_time, has_time) = (
        due_time_opt.unwrap_or_else(|| now.time()),
        due_time_opt.is_some(),
    );
    let due_date_time = if has_date {
        if has_time {
            DueDate::DayTime(NaiveDateTime::new(due_date, due_time))
        } else {
            DueDate::Day(due_date)
        }
    } else if has_time {
        DueDate::DayTime(NaiveDateTime::new(now.date_naive(), due_time))
    } else {
        DueDate::NoDate
    };
    task.due_date = due_date_time;
    Ok(task)
}
#[cfg(test)]
mod test {

    use chrono::{Datelike, Days, NaiveDate, NaiveDateTime, NaiveTime};

    use crate::{
        parser::task::parse_task,
        task::{DueDate, State, Task},
        TasksConfig,
    };
    #[test]
    fn test_parse_task_no_description() {
        let mut input = "- [x] 10/15 task_name #done";
        let config = TasksConfig {
            use_american_format: true,
            ..Default::default()
        };
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let year = chrono::Local::now().year();
        let expected = Task {
            name: "task_name".to_string(),
            description: None,
            tags: Some(vec!["done".to_string()]),
            due_date: DueDate::Day(NaiveDate::from_ymd_opt(year, 10, 15).unwrap()),
            priority: 0,
            state: State::Done,
            line_number: 1,
            ..Default::default()
        };
        assert_eq!(res, expected);
    }

    #[test]
    fn test_parse_task_only_state() {
        let mut input = "- [ ]";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let expected = Task {
            subtasks: vec![],
            name: String::new(),
            description: None,
            tags: None,
            due_date: DueDate::NoDate,
            priority: 0,
            state: State::ToDo,
            line_number: 1,
            filename: String::new(),
            is_today: false,
        };
        assert_eq!(res, expected);
    }
    #[test]
    fn test_parse_task_with_due_date_words() {
        let mut input = "- [ ] today 15:30 task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let expected_date = chrono::Local::now().date_naive();
        let expected_time = NaiveTime::from_hms_opt(15, 30, 0).unwrap();
        let expected_due_date = DueDate::DayTime(NaiveDateTime::new(expected_date, expected_time));
        assert_eq!(res.due_date, expected_due_date);
    }

    #[test]
    fn test_parse_task_with_weekday() {
        let mut input = "- [ ] monday 15:30 task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();

        let now = chrono::Local::now();
        let expected_date = now
            .date_naive()
            .checked_add_days(Days::new(
                8 - u64::from(now.date_naive().weekday().number_from_monday()),
            ))
            .unwrap();
        let expected_time = NaiveTime::from_hms_opt(15, 30, 0).unwrap();
        let expected_due_date = DueDate::DayTime(NaiveDateTime::new(expected_date, expected_time));
        assert_eq!(res.due_date, expected_due_date);
    }

    #[test]
    fn test_parse_task_with_weekday_this() {
        let mut input = "- [ ] this monday 15:30 task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let now = chrono::Local::now();
        let expected_date = now
            .date_naive()
            .checked_add_days(Days::new(
                8 - u64::from(now.date_naive().weekday().number_from_monday()),
            ))
            .unwrap();
        let expected_time = NaiveTime::from_hms_opt(15, 30, 0).unwrap();
        let expected_due_date = DueDate::DayTime(NaiveDateTime::new(expected_date, expected_time));
        assert_eq!(res.due_date, expected_due_date);
    }

    #[test]
    fn test_parse_task_with_weekday_next() {
        let mut input = "- [ ] next monday 15:30 task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let now = chrono::Local::now();
        let expected_date = now
            .date_naive()
            .checked_add_days(Days::new(
                8 - u64::from(now.date_naive().weekday().number_from_monday()),
            ))
            .unwrap();
        let expected_time = NaiveTime::from_hms_opt(15, 30, 0).unwrap();
        let expected_due_date = DueDate::DayTime(NaiveDateTime::new(expected_date, expected_time));
        assert_eq!(res.due_date, expected_due_date);
    }

    #[test]
    fn test_parse_task_without_due_date() {
        let mut input = "- [ ] task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let expected_due_date = DueDate::NoDate;
        assert_eq!(res.due_date, expected_due_date);
    }

    #[test]
    fn test_parse_task_with_invalid_state() {
        let mut input = "- [invalid] task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_err());
    }

    #[test]
    fn test_parse_task_without_state() {
        let mut input = "task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_err());
    }

    #[test]
    fn test_parse_task_with_invalid_priority() {
        let mut input = "- [ ] task_name p-9";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        assert_eq!(res.priority, 0);
    }

    #[test]
    fn test_parse_task_without_name() {
        let mut input = "- [ ]";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        assert_eq!(res.name, ""); // Default name is used when no name is provided
    }
    #[test]
    fn test_parse_task_with_today_flag() {
        let mut input = "- [ ] @t";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        assert!(res.is_today);
    }
}
