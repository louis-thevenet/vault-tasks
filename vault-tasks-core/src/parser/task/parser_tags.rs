use winnow::{combinator::preceded, token::take_while, PResult, Parser};

use super::token::Token;

/// Parses tags of the form "#tag".
pub fn parse_tag(input: &mut &str) -> PResult<Token> {
    let tag = preceded(
        '#',
        take_while(1.., ('_', '0'..='9', 'A'..='Z', 'a'..='z', '0'..='9')),
    )
    .parse_next(input)?;
    Ok(Token::Tag(tag.to_string()))
}

#[cfg(test)]
mod tests {
    use crate::parser::task::{parser_tags::parse_tag, token::Token};

    #[test]
    fn test_parse_tag_sucess() {
        let mut with_tag = "#test";
        assert_eq!(parse_tag(&mut with_tag), Ok(Token::Tag("test".to_string())));
    }
    #[test]
    fn test_parse_tag_symbols() {
        let mut with_tag = "#test_underscore123";
        assert_eq!(
            parse_tag(&mut with_tag),
            Ok(Token::Tag("test_underscore123".to_string()))
        );
    }
    #[test]
    fn test_parse_tag_fail() {
        let mut without_tag = "test";
        assert!(parse_tag(&mut without_tag).is_err());
    }
}
