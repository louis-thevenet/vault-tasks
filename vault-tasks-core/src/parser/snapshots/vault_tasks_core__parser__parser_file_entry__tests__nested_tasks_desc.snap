---
source: vault-tasks-core/src/parser/parser_file_entry.rs
expression: res
---
Test
‾‾‾‾
	1 Header
	‾‾‾‾‾‾‾‾
		❌ t1
		t1
		t1
		t1
		t1
		
			❌ t2
			t2
			t2
			t2
			t2
			
				❌ t3
				t3
				t3
				
					❌ t4
					t4
					t4
